File: leetcode/1172-dinner-plate-stacks/Solution.java
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

class DinnerPlates {
    private final int capacity;
    private final List<List<Integer>> stacks;
    // Min-heap of stack indices that MIGHT have room to push into.
    // Entries can go stale (index removed / stack filled up since being
    // added) -- they're just skipped lazily whenever they surface.
    private final PriorityQueue<Integer> available;

    public DinnerPlates(int capacity) {
        this.capacity = capacity;
        this.stacks = new ArrayList<>();
        this.available = new PriorityQueue<>();
    }

    public void push(int val) {
        while (!available.isEmpty()
                && (available.peek() >= stacks.size() || stacks.get(available.peek()).size() >= capacity)) {
            available.poll();
        }

        int index;
        if (available.isEmpty()) {
            index = stacks.size();
            stacks.add(new ArrayList<>());
        } else {
            index = available.peek();
        }

        stacks.get(index).add(val);

        if (stacks.get(index).size() < capacity) {
            available.offer(index);
        }
    }

    public int pop() {
        while (!stacks.isEmpty() && stacks.get(stacks.size() - 1).isEmpty()) {
            stacks.remove(stacks.size() - 1);
        }

        if (stacks.isEmpty()) {
            return -1;
        }

        return popAtStack(stacks.size() - 1);
    }

    public int popAtStack(int index) {
        if (index >= stacks.size() || stacks.get(index).isEmpty()) {
            return -1;
        }

        List<Integer> stack = stacks.get(index);
        int val = stack.remove(stack.size() - 1);
        available.offer(index);
        return val;
    }
}

/**
 * Your DinnerPlates object will be instantiated and called as such:
 * DinnerPlates obj = new DinnerPlates(capacity);
 * obj.push(val);
 * int param_2 = obj.pop();
 * int param_3 = obj.popAtStack(index);
 */

File: leetcode/1172-dinner-plate-stacks/solution.cpp
#include <vector>
#include <queue>
using namespace std;

class DinnerPlates {
private:
    int capacity;
    vector<vector<int>> stacks;
    // Min-heap of stack indices that MIGHT have room to push into.
    // Entries can go stale (index removed / stack filled up since being
    // added) -- they're just skipped lazily whenever they surface.
    priority_queue<int, vector<int>, greater<int>> available;

public:
    DinnerPlates(int capacity) : capacity(capacity) {}

    void push(int val) {
        while (!available.empty() &&
               (available.top() >= (int)stacks.size() ||
                (int)stacks[available.top()].size() >= capacity)) {
            available.pop();
        }

        int index;
        if (available.empty()) {
            index = stacks.size();
            stacks.push_back({});
        } else {
            index = available.top();
        }

        stacks[index].push_back(val);

        if ((int)stacks[index].size() < capacity) {
            available.push(index);
        }
    }

    int pop() {
        while (!stacks.empty() && stacks.back().empty()) {
            stacks.pop_back();
        }

        if (stacks.empty()) {
            return -1;
        }

        return popAtStack(stacks.size() - 1);
    }

    int popAtStack(int index) {
        if (index >= (int)stacks.size() || stacks[index].empty()) {
            return -1;
        }

        int val = stacks[index].back();
        stacks[index].pop_back();
        available.push(index);
        return val;
    }
};

/**
 * Your DinnerPlates object will be instantiated and called as such:
 * DinnerPlates* obj = new DinnerPlates(capacity);
 * obj->push(val);
 * int param_2 = obj->pop();
 * int param_3 = obj->popAtStack(index);
 */
